### ISSUE TITLE
Silent error in handling of univariate case

### ISSUE THREAD

**Issue by johnpjust** [nicola-decao/BNAF#2]
The forward pass of the BNAF class should be changed to the following (or similar) to properly handle the univariate case or calling model will reduce the gradients (grad) to a single number when it should have a first dim size of batch_size.  I confirmed the following will work though.

       ```
        grad = grad.squeeze()
        reduce_sum = len(grad.shape) > 1

        if reduce_sum:
            if self.res == 'normal':
                return inputs + outputs, torch.nn.functional.softplus(grad.squeeze()).sum(-1)
            elif self.res == 'gated':
                return self.gate.sigmoid() * outputs + (1 - self.gate.sigmoid()) * inputs, \
                       (torch.nn.functional.softplus(grad.squeeze() + self.gate) - \
                        torch.nn.functional.softplus(self.gate)).sum(-1)
            else:
                return outputs, grad.squeeze().sum(-1)
        else:
            if self.res == 'normal':
                return inputs + outputs, torch.nn.functional.softplus(grad)
            elif self.res == 'gated':
                return self.gate.sigmoid() * outputs + (1 - self.gate.sigmoid()) * inputs, \
                       (torch.nn.functional.softplus(grad + self.gate) - \
                        torch.nn.functional.softplus(self.gate))
            else:
                return outputs, grad
```

**Comment by nicola-decao** [comment]
Thanks! I'll include that.

**Comment by nicola-decao** [comment]
With an univariate case you will have `[batch_size, 1]` does shouldn't break things.
